### PR TITLE
Add extension mechanism

### PR DESCRIPTION
**Pull request by fpagnoux**
Depends on #8

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,20 +6,22 @@ This repository eases the deployment of a production [OpenFisca](http://openfisc
...
 
-OpenFisca is comprised of four different parts:
+OpenFisca is comprised of four different parts, to which you can add optional [extensions](http://doc.openfisca.fr/contribute/extensions.html):
 
 - [`openfisca-core`](https://github.com/openfisca/openfisca-core)
 - [`openfisca-france`](https://github.com/openfisca/openfisca-france)
 - [`openfisca-parsers`](https://github.com/openfisca/openfisca-parsers)
 - [`openfisca-web-api`](https://github.com/openfisca/openfisca-web-api)
 
+Example of extension: [`openfisca-paris`](https://github.com/sgmap/openfisca-paris)
```
Comment (MattiSG) [inline README.md]: Ponctuation.

File: README.md [hunk 2]
```diff
@@ -27,6 +29,8 @@ This repository references specific versions of each part of OpenFisca, and exec
 The tests come from the few tests that are available within each OpenFisca module, and from client applications that have created their own functional tests such as [`mes-aides`](https://github.com/sgmap/mes-aides-api). We aim at adding as many tests as possible to ensure maximal coverage.
 
 
+
+
```
Comment (MattiSG) [inline README.md]: ?

File: README.md [hunk 3]
```diff
@@ -35,8 +39,15 @@ Usage
...
 
+
 ### Run
 
 ```
 git clone https://github.com/sgmap/openfisca.git && openfisca/update.sh
 ```
+
+### Extensions
+To add an extension to this repository :
```
Comment (MattiSG) [inline README.md]: No leading space.

File: update.sh [hunk 4]
```diff
@@ -5,6 +5,11 @@ set -ex
 
 cd `dirname $0`
 
+if [[ ! $VIRTUAL_ENV ]]
+then
+	user_option='--user'
```
Comment (MattiSG) [inline update.sh]: Maybe add a comment like "sandbox the install to the virtualenv or, if not available, to the user"?

File: update.sh [hunk 5]
```diff
@@ -14,5 +19,23 @@ fi
 git submodule sync
 git submodule update --init --recursive
 
-git submodule foreach python setup.py develop --user
-git submodule foreach pip install --user --editable .
+for submodule in core web-api france parsers
+do
+	cd openfisca/$submodule/
+	python setup.py develop $user_option
+	pip install $user_option  --editable .
```
Comment (MattiSG) [inline update.sh]: Double space.

File: update.sh [hunk 6]
```diff
@@ -14,5 +19,23 @@ fi
...
+	cd ../..
```
Comment (MattiSG) [inline update.sh]: > `cd -` is a simpler and more robust equivalent.

File: update.sh [hunk 7]
```diff
@@ -14,5 +19,23 @@ fi
...
+done
+
+for old_symlink_extension in $(find openfisca/france/openfisca_france/model/extensions -type l)
+do
+	rm $old_symlink_extension
+done
+
+for extension in $(ls  -d openfisca/extensions/*)
```
Comment (MattiSG) [inline update.sh]: Double space.

Comment (MattiSG) [inline update.sh]: Is the `*` necessary?

File: update.sh [hunk 8]
```diff
@@ -14,5 +19,23 @@ fi
...
+do
+	if [[  $(git submodule | grep $extension) ]]
```
Comment (MattiSG) [inline update.sh]: Why `test` the result of a subshell execution rather than simply checking the exit value?

File: update.sh [hunk 9]
```diff
@@ -14,5 +19,23 @@ fi
...
+	then
+		ln -s $(pwd)/$extension openfisca/france/openfisca_france/model/extensions/
+	fi
+done
```
Comment (MattiSG) [inline update.sh]: Wouldn't it be simpler to have a single iteration loop over the submodules, check whether they are in the `openfisca/extensions` directory and branch to either `pip install` or `ln -s`? You can even put each treatment in its own function to make it clearer.

**Comment by MattiSG** [comment]
Wouldn't it be simpler to have a single iteration loop over the submodules, check whether they are in the `openfisca/extensions` directory and branch to either `pip install` or `ln -s`? You can even put each treatment in its own function to make it clearer.

I also draw your attention to this nicety of `git submodule foreach`:

>  The command has access to the variables $name, $path,  $sha1 and $toplevel: $name is the name of the relevant submodule section in .gitmodules, $path is the name of the submodule directory relative to the super project.

:)

File: update.sh (hunk 4 above)
Comment (fpagnoux) [inline update.sh]: I learnt a new verb, thanks :wink:

File: update.sh (hunk 7 above)
Comment (fpagnoux) [inline update.sh]: Yep

``` sh
ls -d openfisca/extensions/*
>>>  openfisca/extensions/paris
ls -d openfisca/extensions/
>>> openfisca/extensions/
```

File: update.sh (hunk 8 above)
Comment (fpagnoux) [inline update.sh]: Do this mean I don't need the $( ) ?

File: README.md [hunk 10]
```diff
@@ -40,3 +42,9 @@ Usage
 ```
 git clone https://github.com/sgmap/openfisca.git && openfisca/update.sh
 ```
+
+### Extensions
+To add an extension to this repository:
+```sh
+git submodule add {extension_repository} openfisca/extensions/{extension_name}
```
Comment (MattiSG) [inline README.md]: > When in shell, use `$` for example templates  ;)

File: install_submodule.sh [hunk 11]
```diff
@@ -0,0 +1,16 @@
+path=$1
+toplevel=$2
+
+# Sandbox the install to the virtualenv, or, if not available, to the user.
+if [[ ! $VIRTUAL_ENV ]]
+then
+	user_option='--user'
+fi
+
+if echo $path | grep openfisca/extensions
```
Comment (MattiSG) [inline install_submodule.sh]: This will output the path. Is it expected? If no, add `-q`.

**Comment by MattiSG** [comment]
:+1: GTM

**Pull request merged**